### DISCUSSION TITLE
Uncaught Error: Class "Shlinkio\Shlink\Core\Config\EnvVars" not found

### DISCUSSION THREAD

**Discussion by danielsuareztech** [shlinkio/shlink#1918] [Help wanted]
### Shlink version

3.6.4

### PHP version

8.1.25

### How do you serve Shlink

Self-hosted nginx

### Database engine

MySQL

### Database version

8.0.35

### Summary

I've followed all the installation instructions, but I'm struggling to manually create the first API key.

The command being executed is:
`sudo /var/www/html/shlink/bin/cli api-key:generate`

The error message I'm getting is:

**PHP Fatal error:  Uncaught Error: Class "Shlinkio\Shlink\Core\Config\EnvVars" not found in /var/www/html/shlink/config/container.php:26
Stack trace:
#0 /var/www/html/shlink/config/cli-app.php(10): include()
#1 /var/www/html/shlink/config/cli-app.php(12): {closure}()
#2 /var/www/html/shlink/bin/cli(9): require('...')
#3 {main}
  thrown in /var/www/html/shlink/config/container.php on line 26**

I've installed Shlink with the provided dist file for PHP 8.1 and i'm using the following global Environment Variables defned in \etc\environment.

**DB_DRIVER="mysql"
DB_HOST="192.168.1.2"
DB_PORT="3306"
DB_NAME="shlink"
DB_USER="shlink_rw"
DB_PASSWORD="MyStrongPassword"
TIMEZONE="America/Toronto"
DEFAULT_DOMAIN="short.url"
IS_HTTPS_ENABLED=true
BASE_PATH="/var/www/html/shlink"
GEOLITE_LICENSE_KEY="MyGeoliteLicense"**

I'm far from being a knowledgeable linux user, so, I might be making any silly mistake but I'm not able to figure out what it could be.

I'm using the environments variables approach because I'm facing different issues when attempting to install shlink-installer but that's for a different discussion.

Thanks,
Regards

**Comment by acelaya** [comment]
This is a very strange error to happen. If this was an actual general problem, many people would be complaining.

Nonetheless, I have just downloaded the PHP 8.1 dist file for Shlink 3.6.4 to make sure the autoloader was correct, and it is.

I then spin up a docker container with PHP 8.1.25, downloaded the dist file there, and run it. Everything worked as expected.

I have tried both running the command with a relative path, from Shlink's root folder, and also using an absolute path. Both worked.

Things I think could be happening:

1. If I had to bet, I would say you downloaded the source code, not a dist file. Source code is not pre-built, and the autoloader is missing (among other things), which would cause this error.
  If you clicked here, where it says "Download ZIP", that's the raw source code.
  ![image](https://github.com/shlinkio/shlink/assets/2719332/a9b6e443-22d9-437a-8992-281319f77c2f)
  Dist files can be found [here](https://github.com/shlinkio/shlink/releases/tag/v3.6.4):
  ![image](https://github.com/shlinkio/shlink/assets/2719332/1ef7cb49-fbb7-401b-a506-babb2a119688)
  Below that there's, again, packed source files.
2. Perhaps using `sudo` has some influence on how paths are resolved, but I couldn't verify this inside the docker container.
  Shlink does not need to be run as superuser, so try running the command without `sudo`.

Unrelated issues I see:

* When you pass configuration via env vars, there's an additional required step you need to run once at the beginning. `vendor/bin/shlink-installer init`.
  That will set-up everything for Shlink to work.
  Skipping this would cause errors regarding missing database tables and such, but not the error reported here.
* The `BASE_PATH` env var does not represent the filesystem directory where Shlink is located, but the URL path prefix in case you are not serving Shlink from the root of your domain.
  Some people use the same domain for other things and this comes handy.
  Passing `/var/www/html/shlink` means Shlink will expect all short URLs to be in the form of `https://short.url/var/www/html/shlink/{short-code}`, and API requests will not work.

**Comment by danielsuareztech** [comment]
Thanks for taking the time to reply.

I'm not sure how it got fixed but I followed the following steps:

1) Deployed composer by following the steps [here](https://www.digitalocean.com/community/tutorials/how-to-install-and-use-composer-on-ubuntu-20-04)
1) Deleted everything
2) Downloaded everything again
3) Changed ownership recursively
4) Disabled environment variables (commented all the variables in /etc/environment)
5) Ran the installer with sudo -u www-data php vendor/bin/shlink-installer install

The installation ran successfully, and I got my first API Key automatically created, making me think the API is now working fine.

I'm now having what looks like a PHP processing error but I will open a different discussion thread

So, it looks like the problem was behind the keyboard: me.

Thank you so much!

**Reply by acelaya** [reply]
Just to reiterate on what was mentioned, make sure you are downloading a dist file, and following these instructions https://shlink.io/documentation/install-dist-file/#installation-steps

Installing composer is not part of those instructions. You do NOT need composer to run Shlink. Composer is a tool used to develop Shlink.

If you are following something that makes you think you need composer, you are not following the right process.
